File: twilio-php/phonemenu/PhoneMenu.java
package com.twilio;

import java.util.HashMap;
import java.util.Map;

import com.twilio.sdk.verbs.*;

public class PhoneMenu {
	
	static String BASEURL="http://www.enter_your_domain_here.com:8080/phonemenu/phonemenu.jsp";
	
	public static void main(String[] args) {
		System.out.println(PhoneMenu.getTwiML(null,null));
	}

	public static String getTwiML(String node, String digits){
		// Parse digits; represent lack of digits as -1
        int index;
        try
        	{index = Integer.parseInt(digits);}
        catch (NumberFormatException nfe)
        	{ index=-1;}

		// @start snippet
		/* Define Menu */
		String[] defaultMenu = {"receptionist","hours", "location", "duck"};
		String[] locationMenu = {"receptionist","east-bay", "san-jose"};

        Map<String,String[]> menuDefinition = new HashMap<String,String[]>();
        menuDefinition.put("default", defaultMenu);
        menuDefinition.put("location", locationMenu);

        String destination;
		/* Check to make sure index is valid */
		if(index>=0 && node!=null && menuDefinition.get(node).length >= index)
			destination = menuDefinition.get(node)[index];
		else
			destination = null;
		// @end snippet

		TwiMLResponse response = new TwiMLResponse();

        try {
	
			// @start snippet
			if (destination=="receptionist"){
		        Say say = new Say("Please wait while we connect you");
		        Dial dial = new Dial("8583829141");
		        response.append(say);
		        response.append(dial);
			} else if (destination=="hours"){
				Say say = new Say("Initech is open Monday through Friday, 9am to 5pm, Saturday, 10am to 3pm and closed on Sundays.");
				response.append(say);
			} else if (destination=="location"){
				Say say = new Say("Initech is located at 101 4th St in San Francisco California");
				response.append(say);
			} else if (destination=="duck"){
				Play play = new Play("duck.mp3");
				response.append(play);
			} else if (destination=="receptionist"){
				Say say = new Say("Initech is open Monday through Friday, 9am to 5pm, Saturday, 10am to 3pm and closed on Sundays.");
				response.append(say);
			} else if (destination=="east-bay"){
				Say say = new Say("Take BART towards San Francisco / Milbrae. Get off on Powell Street. Walk a block down 4th street.");
				response.append(say);
			} else if (destination=="san-jose"){
				Say say = new Say("Take Cal Train to the Milbrae BART station. Take any Bart train to Powell Street.");
				response.append(say);
			} else { 
				// no destination
				Gather gather = new Gather();
				gather.setAction(BASEURL + "?node=default");
				gather.setNumDigits(1);
				Say say = new Say("Hello and welcome to the Initech Phone Menu. For business hours, press 1.  For directions, press 2.  To hear a duck quack, press 3 To speak to a receptionist, press 0");
				gather.append(say);
				response.append(gather);
			}            
			
	    	if (destination=="location"){
	    		Gather gather = new Gather();
	    		gather.setAction(BASEURL + "?node=location");
	    		gather.setNumDigits(1);
	    		response.append(gather);
	    		gather.append (new Say ("For directions from the East Bay, press 1. For directions from San Jose, press 2."));
	    	}
	    	
			// @end snippet
			// @start snippet
	    	if (destination!=null && destination!="receptionist") {
				Pause pause = new Pause();
				pause.setLength(2);
				response.append(pause);
				Say menuSay = new Say("Main Menu");
				response.append(menuSay);
				Redirect redirect = new Redirect(BASEURL);
				response.append(redirect);
			}
			// @end snippet

        } catch (TwiMLException e) {
            e.printStackTrace();
        }

        return(response.toXML());
      }
}

File: twilio-php/phonemenu/phonemenu.php
<?php

// @start snippet
/* Define Menu */
$web = array();
$web['default'] = array('receptionist','hours', 'location', 'duck');
$web['location'] = array('receptionist','east-bay', 'san-jose', 'marin');
$web['days'] = array('monday','tuesday','wednesday','thursday','friday','default');
$web['choice_monday'] = array('','monday');
$web['choice_tuesday'] = array('tuesday');
$web['choice_wednesday'] = array('wednesday');
$web['choice_thursday'] = array('thursday');
$web['choice_friday'] = array('friday');
$web['choice_monday_time'] = array('','monday_time');
$web['choices'] = array('1','2');

/* Get the menu node, index, and url */
$node = $_REQUEST['node'];
$index = (int) $_REQUEST['Digits'];
$url = 'http://'.dirname($_SERVER["SERVER_NAME"].$_SERVER['PHP_SELF']).'/phonemenu.php';
//$day = $_REQUEST['day'];
/* get the day and the choice */
//$day_index  = array_search($_REQUEST['day'], $web['days']);

/* Check to make sure index is valid */
if(isset($web[$node]) || count($web[$node]) >= $index && !is_null($_REQUEST['Digits'])){
    $destination = $web[$node][$index];
}
else{
    $destination = NULL;   
}
// @end snippet

// @start snippet
/* Render TwiML */
header("content-type: text/xml");
echo "<?xml version=\"1.0\" encoding=\"UTF-8\"?><Response>\n";
switch($destination) {
    case 'hours': ?>
        <Say>Initech is open Monday through Friday, 9am to 5pm</Say>
        <Say>Saturday, 10am to 3pm and closed on Sundays</Say>
        <?php break;
    case 'location': ?>
        <Say>Initech is located at 101 4th St in San Francisco California</Say>
        <Gather action="<?php echo 'http://' . dirname($_SERVER["SERVER_NAME"] .  $_SERVER['PHP_SELF']) . '/phonemenu.php?node=location'; ?>" numDigits="1">
            <Say>For directions from the East Bay, press 1</Say>
            <Say>For directions from San Jose, press 2</Say>
        </Gather>
    <?php break;

    case 'monday'; ?>
        <Gather action="<?php echo 'http://' . dirname($_SERVER["SERVER_NAME"] .  $_SERVER['PHP_SELF']) . '/phonemenu.php?node=choice_monday_time'; ?>" numDigits="1">
                <Say>If you are free on Monday morning between 8 to 10 press 1 else press 2</Say>
        </Gather>
        <?php break;
    case 'monday_time'; ?>
        <Say> Great, we have fixed a session for you on Monday with a Raise Voice teacher. We will send you a reminder text before the session </Say>
        
        <?php break;
    case 'tuesday'; ?>
     <Gather action="<?php echo 'http://' . dirname($_SERVER["SERVER_NAME"] .  $_SERVER['PHP_SELF']) . '/phonemenu.php?node=choice_tuesday'; ?>" numDigits="1">
               
                <Say>If you are free on Tuesday morning between 8 to 10 press 1 else press 2</Say>
    </Gather>
        <?php break;
    case 'wednesday'; ?>
     <Gather action="<?php echo 'http://' . dirname($_SERVER["SERVER_NAME"] .  $_SERVER['PHP_SELF']) . '/phonemenu.php?node=choice_wednesday'; ?>" numDigits="1">

                <Say>If you are free on Wednesday morning between 8 to 10 press 1 else press 2</Say>
    </Gather>
    <?php break;    
    case 'thursday'; ?>
     <Gather action="<?php echo 'http://' . dirname($_SERVER["SERVER_NAME"] .  $_SERVER['PHP_SELF']) . '/phonemenu.php?node=choice_thursday'; ?>" numDigits="1">
                <Say>If you are free on Thursday morning between 8 to 10 press 1 else press 2</Say>
    </Gather>
        <?php break;
    case 'friday'; ?>
     <Gather action="<?php echo 'http://' . dirname($_SERVER["SERVER_NAME"] .  $_SERVER['PHP_SELF']) . '/phonemenu.php?node=choice_friday'; ?>" numDigits="1">
                <Say>If you are free on Friday morning between 8 to 10 press 1 else press 2</Say>
    </Gather>
        <?php break;
    default: ?>
        <Gather action="<?php echo 'http://' . dirname($_SERVER["SERVER_NAME"] .  $_SERVER['PHP_SELF']) . '/phonemenu.php?node=choice_monday'; ?>" numDigits="1">
            <Say>Hello and welcome to the Raise Voice Menu</Say>
            <Say>Lets complete our registration process</Say>
            <Say>Press 1 to start telling us the days else press 2 to hang up</Say>
        </Gather>
        <?php
    break;
}
// @end snippet

// @start snippet
if($destination && $destination != 'receptionist') { ?>
    <Pause/>
   <Say>Main Menu</Say>
    <Redirect><?php echo 'http://' . dirname($_SERVER["SERVER_NAME"] .  $_SERVER['PHP_SELF']) . '/phonemenu.php' ?></Redirect>
<?php }
// @end snippet

?>

</Response>

